### PR TITLE
Improve yast2_apparmor.pm against low performance

### PR DESCRIPTION
**Pull request by Zaoliang**
- add workaround agains timeout issue by using wait_screen_chang and increasing timeout value
- add required needle checks for file and directory permission
- please see ticket: https://progress.opensuse.org/issues/20202
- please see reference tests: 
openSUSE Leap: http://e13.suse.de/tests/3643#step/yast2_apparmor
openSUSE TW: http://e13.suse.de/tests/3649#step/yast2_apparmor
SLES 12 sp3: http://e13.suse.de/tests/3651#step/yast2_apparmor

- please also check needle PR: 
SLES 12 SP3: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/429
openSUSE:  https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/243

### CODE REVIEW

**Comment by SergioAtSUSE** [comment]
The commit message could be improved to something like: Improve yast2_apparmor against low performance

Is there a ticket that you could refer here?
In case that not, is there a job result on o3 or osd where the original problem can be seen?

**Comment by Zaoliang** [comment]
commit message got updated.

File: tests/console/yast2_apparmor.pm [hunk 1]
```diff
@@ -24,11 +24,10 @@ sub run {
     assert_script_run("/usr/bin/zypper -n -q in yast2-apparmor");
 
     # start apparmor configuration
-    script_run("yast2 apparmor; echo yast2-apparmor-status-\$? > /dev/$serialdev", 0);
-
+    # script_run("yast2 apparmor; echo yast2-apparmor-status-\$? > /dev/$serialdev", 0);
```
Comment (SergioAtSUSE) [inline tests/console/yast2_apparmor.pm]: Instead of leaving this line commented, I would leave a comment explaining why this change is necessary.

File: tests/console/yast2_apparmor.pm [hunk 2]
```diff
@@ -38,123 +37,98 @@ sub run {
...
+    wait_screen_change { send_key 'alt-b' };
 
     # close apparmor configuration
-    send_key 'alt-d';
-    wait_still_screen(1);
+    wait_screen_change { send_key 'alt-d' };
     assert_script_run("systemctl show -p ActiveState apparmor.service | grep ActiveState=active");
 
     # part 2: start apparmor configuration again
-    script_run("yast2 apparmor; echo yast2-apparmor-status-\$? > /dev/$serialdev", 0);
```
Comment (SergioAtSUSE) [inline tests/console/yast2_apparmor.pm]: Same here: Instead of leaving this line commented, I would leave a comment explaining why this change is necessary.

**Comment by okurz** [comment]
please also update PR subject

File: tests/console/yast2_apparmor.pm (hunk 2 above)
Comment (Zaoliang) [inline tests/console/yast2_apparmor.pm]: The problem with wait_serial("yast2-apparmor-status-0", 60) || die "'yast2 apparmor' didn't finish"; 
We need to start apparmor for couple of times during testing and it has sometimes timeout issue for checking it's execution status. And it just failed sometimes. Please advice if you#re not agree with this.

Comment (SergioAtSUSE) [inline tests/console/yast2_apparmor.pm]: That explanation is good for me, but I also mean that you should add that sentence as comment in the code:

```
# Using type_string instead of script_run because of timeout problems with wait_serial: poo#20202
```

File: tests/console/yast2_apparmor.pm (hunk 1 above)
Comment (SergioAtSUSE) [inline tests/console/yast2_apparmor.pm]: Same here:
```
# Using type_string instead of script_run because of timeout problems with wait_serial: poo#20202
```

File: tests/console/yast2_apparmor.pm (hunk 2 above)
Comment (Zaoliang) [inline tests/console/yast2_apparmor.pm]: okay, just added this in commit message, please check, thanks!

Comment (Zaoliang) [inline tests/console/yast2_apparmor.pm]: and add comment for this change before type_string

File: tests/console/yast2_apparmor.pm [hunk 3]
```diff
@@ -38,123 +40,98 @@ sub run {
     }
     assert_screen 'yast2_apparmor_enabled';
     # part 1: open profile mode configuration and check toggle/show all profiles
-    send_key 'alt-n';
+    wait_screen_change { send_key 'alt-n' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: you should not need a wait_screen_change before an assert_screen because that offers a synchronisation point already. No point in waiting twice

File: tests/console/yast2_apparmor.pm [hunk 4]
```diff
@@ -38,123 +40,98 @@ sub run {
...
     assert_screen 'yast2_apparmor_profile_mode_configuration';
-    send_key 'alt-o';
-    wait_still_screen(1);
+    wait_screen_change { send_key 'alt-o' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: same here

File: tests/console/yast2_apparmor.pm [hunk 5]
```diff
@@ -38,123 +40,98 @@ sub run {
...
     assert_screen 'yast2_apparmor_profile_mode_configuration_show_all';
-    wait_still_screen(1);
-    send_key 'tab';
-    wait_still_screen(1);
-    send_key 'down';
-    wait_still_screen(1);
-    send_key 'alt-t';
+    wait_screen_change { send_key 'tab' };
+    wait_screen_change { send_key 'down' };
+    wait_screen_change { send_key 'alt-t' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: same here

File: tests/console/yast2_apparmor.pm (hunk 2 above)
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: sorry, but I don't agree with that. `script_run(…, 0)` is equivalent to `type_string "…\n"` so that should not matter. I don't see a point why you would not want to test the exit code. That should be preserved. Please change back to the previous state. If it fails sometimes because it takes longer and you think that the additional waiting time would be acceptable then please just add more to the timeout on the corresponding wait_serial call.

**Comment by Zaoliang** [comment]
okay, then revert this change and re-test the code.

File: tests/console/yast2_apparmor.pm [hunk 6]
```diff
@@ -24,11 +24,9 @@ sub run {
     assert_script_run("/usr/bin/zypper -n -q in yast2-apparmor");
 
     # start apparmor configuration
-    script_run("yast2 apparmor; echo yast2-apparmor-status-\$? > /dev/$serialdev", 0);
-
     # check Apparmor Configuration is opened
+    script_run("yast2 apparmor; echo yast2-apparmor-status-\$? > /dev/$serialdev", 0);
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: that should go back to above, not here because it does not fit to the comment anymore

File: tests/console/yast2_apparmor.pm (hunk 3 above)
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: comment still applies here

File: tests/console/yast2_apparmor.pm [hunk 7]
```diff
@@ -25,38 +25,30 @@ sub run {
...
+    if (match_has_tag 'yast2_apparmor_disabled') {
+        send_key 'alt-e';
     }
     assert_screen 'yast2_apparmor_enabled';
     # part 1: open profile mode configuration and check toggle/show all profiles
     send_key 'alt-n';
     assert_screen 'yast2_apparmor_profile_mode_configuration';
-    send_key 'alt-o';
-    wait_still_screen(1);
+    wait_screen_change { send_key 'alt-o' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: wait_screen_change before assert_screen -> redundant

File: tests/console/yast2_apparmor.pm [hunk 8]
```diff
@@ -66,95 +58,79 @@ sub run {
...
     assert_screen 'yast2_apparmor_configuration_manage_existing_profiles_edit_add';
-    send_key 'alt-i';
-    wait_still_screen(1);
-    send_key 'alt-a';
-    wait_still_screen(1);
-    send_key 'alt-f';
-    wait_still_screen(1);
+    wait_screen_change { send_key 'alt-i' };
+    wait_screen_change { send_key 'alt-a' };
+    wait_screen_change { send_key 'alt-f' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: same here

File: tests/console/yast2_apparmor.pm [hunk 9]
```diff
@@ -66,95 +58,79 @@ sub run {
...
-    send_key 'spc';
-    send_key 'alt-o';
-    wait_still_screen(1);
+    wait_screen_change { send_key 'alt-p' };
+    wait_screen_change { send_key 'down' };
+    wait_screen_change { send_key 'up' };
+    wait_screen_change { send_key 'spc' };
+    wait_screen_change { send_key 'down' };
+    wait_screen_change { send_key 'spc' };
+    wait_screen_change { send_key 'alt-o' };
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: this is a huge list of actions with no assert_screen in between means that if there is something going wrong it would be very hard to find the key press that went wrong. I suggest to make some intermediate steps assert_screen instead.

@rwx788 ?

File: tests/console/yast2_apparmor.pm [hunk 10]
```diff
@@ -66,95 +58,79 @@ sub run {
...
     send_key 'alt-d';
 
     # confirm to save changes to the profile
     assert_screen 'yast2_apparmor_configuration_manage_existing_profiles_edit_file_changed';
     send_key 'alt-y';
+    # we had problem here after send_key, use wait_idle to ensure the following
+    # send_key to be executed successfully
+    wait_idle;
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: please avoid wait_idle

File: tests/console/yast2_apparmor.pm [hunk 11]
```diff
@@ -66,95 +58,79 @@ sub run {
...
     type_string '/new_profile';
-    wait_still_screen(1);
-    send_key 'alt-o';
-    wait_still_screen(2);
+    wait_screen_change { send_key 'alt-o' };
 
     # Add entry for new profile
+    # The main dialog shows up for a second before AppArmor Profile Dialog
+    # It waits for AppArmor Profile Dialog showing up
+    wait_idle;
```
Comment (okurz) [inline tests/console/yast2_apparmor.pm]: no

File: tests/console/yast2_apparmor.pm (hunk 9 above)
Comment (rwx788) [inline tests/console/yast2_apparmor.pm]: Totally agreed. We put some timeouts, but it doesn't give a guarantee that we got reaction on key press. It's one of the main issues with yast2_ui tests apart from other sources of instability.

**Comment by okurz** [comment]
both needle PRs merged
